### PR TITLE
Fix ASP.NET Core Traces Instrumentation for .NET7+

### LINKED ISSUE #3212: Activity.Kind improperly set and Activity.ActivitySourceName missing when using Auto-Instrumentation for incoming requests
# Bug Report

## Symptom

**Describe the bug**
I believe the `Activity.Kind` being produced from my traces are not properly set to `Server` as expected and and `Activity.ActivitySourceName` is empty when using auto-instrumentation on a self-contained simple .NET 8.0 web app. 

Because the Kind is not Server I _believe_ this is affecting my service mapping in both AWS X-RAY/Grafana. 

**Expected behavior**
I expect `Activity.Kind` to be `Server` for incoming HTTP requests and to have a non-empty `ActivitySourceName`. Looking at the getting started docs I expect something like:
```
Activity.ActivitySourceName: OpenTelemetry.Instrumentation.AspNetCore
Activity.DisplayName:        /rolldice
Activity.Kind:               Server
```

**Actual**
```
Activity.TraceId:            8a412ea70c2454ea35c60848131c7a91
Activity.SpanId:             d6409440d1c1b04e
Activity.TraceFlags:         Recorded
Activity.ActivitySourceName:
Activity.DisplayName:        GET /rolldice/{player?}
Activity.Kind:               Internal
Activity.StartTime:          2024-01-03T18:49:32.9035202Z
Activity.Duration:           00:00:00.0388405
Activity.Tags:
    net.host.name: localhost
    net.host.port: 8080
    http.method: GET
    http.scheme: http
    http.target: /rolldice
    http.url: http://localhost:8080/rolldice
    http.flavor: 1.1
    http.user_agent: curl/8.1.2
    http.route: /rolldice/{player?}
    http.status_code: 200
Resource associated with Activity:
    container.id: 3f41cd49cfffe4008f6e4a58fff32adf7f7bd4275a7a1d6dc902a9fd2718f5fb
    telemetry.distro.name: opentelemetry-dotnet-instrumentation
    telemetry.distro.version: 1.2.0
    telemetry.sdk.name: opentelemetry
    telemetry.sdk.language: dotnet
    telemetry.sdk.version: 1.6.0
    service.name: hello-world
```

**Runtime environment (please complete the following information):**

- OpenTelemetry Automatic Instrumentation version: 1.2.0
- OS: Ubuntu 22.04
- .NET version: 8

## Reproduce

 1. Create a simple app
 2. `dotnet new webapi`
 3. `dotnet add package OpenTelemetry.AutoInstrumentation --prerelease`
 4. Modify code from the example in the Getting Started docs: https://opentelemetry.io/docs/instrumentation/net/automatic/getting-started/
 5. Build/Run the application using hardcoded environment variables derived from the `instrument.sh` that would be exported from `instrument.sh`

```
FROM mcr.microsoft.com/dotnet/sdk:8.0-jammy AS build
WORKDIR /source

COPY src/hello-world/ .

RUN dotnet publish --runtime linux-x64 -v q -c Release -o /app --self-contained

FROM mcr.microsoft.com/dotnet/runtime:8.0-jammy

ENV OTEL_TRACES_EXPORTER=none
ENV OTEL_METRICS_EXPORTER=none
ENV OTEL_LOGS_EXPORTER=none
ENV OTEL_DOTNET_AUTO_TRACES_CONSOLE_EXPORTER_ENABLED=true
ENV OTEL_SERVICE_NAME=hello-world

ENV ASPNETCORE_HOSTINGSTARTUPASSEMBLIES=OpenTelemetry.AutoInstrumentation.AspNetCoreBootstrapper
ENV CORECLR_ENABLE_PROFILING=1
ENV CORECLR_PROFILER={918728DD-259F-4A6A-AC2B-B85E1B658318}
ENV CORECLR_PROFILER_PATH=/app/OpenTelemetry.AutoInstrumentation.Native.so
ENV DOTNET_STARTUP_HOOKS=/app/OpenTelemetry.AutoInstrumentation.StartupHook.dll
ENV OTEL_DOTNET_AUTO_HOME=/app
ENV OTEL_DOTNET_AUTO_RULE_ENGINE_ENABLED=false

WORKDIR /app
COPY --from=build /app .
ENTRYPOINT ["./hello-world"]
```
 8.  Hit `/rolldice` route and observe metric telemetry

```
otel-dotnet-auto-hello-world-1-20240103.log
[2024-01-03T19:08:41.8672642Z] [Information] OpenTelemetry tracer initialized.
[2024-01-03T19:08:41.8849735Z] [Information] EventSource=OpenTelemetry-AutoInstrumentation, Message=OpenTelemetry tracer initialized.
[2024-01-03T19:08:41.9219648Z] [Information] OpenTelemetry meter initialized.
[2024-01-03T19:08:41.9241148Z] [Information] EventSource=OpenTelemetry-AutoInstrumentation, Message=OpenTelemetry meter initialized.
[2024-01-03T19:08:41.9858280Z] [Information] The profiler has been initialized with 17 definitions for FFAFA5168C4F4718B40CA8788875C2DA.
[2024-01-03T19:08:41.9878884Z] [Information] EventSource=OpenTelemetry-AutoInstrumentation, Message=The profiler has been initialized with 17 definitions for FFAFA5168C4F4718B40CA8788875C2DA.
[2024-01-03T19:08:41.9913448Z] [Information] The profiler has been initialized with 0 derived definitions.
[2024-01-03T19:08:41.9933661Z] [Information] EventSource=OpenTelemetry-AutoInstrumentation, Message=The profiler has been initialized with 0 derived definitions.
[2024-01-03T19:08:41.9972930Z] [Information] EventSource=OpenTelemetry-AutoInstrumentation-StartupHook, Message=StartupHook initialized successfully!
[2024-01-03T19:08:42.0341182Z] [Information] EventSource=OpenTelemetry-AutoInstrumentation, Message=Logs: Loaded AddOpenTelemetry from LoggingBuilder.
[2024-01-03T19:08:42.1282065Z] [Information] EventSource=OpenTelemetry-AutoInstrumentation, Message=BootstrapperHostingStartup loaded for application with name hello-world.
```

```
otel-dotnet-auto-hello-world-1-AspNetCoreBootstrapper-20240103.log
[2024-01-03T19:08:42.1210927Z] [Information] BootstrapperHostingStartup loaded for application with name hello-world.
```

```
otel-dotnet-auto-hello-world-1-Loader-20240103.log
[2024-01-03T19:08:41.4389461Z] [Information] Managed Loader TryLoadManagedAssembly()
```

```
otel-dotnet-auto-hello-world-1-StartupHook-20240103.log
[2024-01-03T19:08:41.2306398Z] [Information] Rule Engine: MinSupportedFrameworkRule evaluation success.
[2024-01-03T19:08:41.3866140Z] [Information] OTEL_DOTNET_AUTO_RULE_ENGINE_ENABLED is set to false, skipping rule engine validation.
[2024-01-03T19:08:41.3886278Z] [Information] Initialization.
[2024-01-03T19:08:41.9952490Z] [Information] StartupHook initialized successfully!
```

```
otel-dotnet-auto-native-hello-world-1.log
[2024-01-03T19:08:40.846585637Z] [1|1] [info] OpenTelemetry CLR Profiler 1.2.0 on Linux (amd64)
[2024-01-03T19:08:40.846876842Z] [1|1] [info] JIT Inlining is enabled.
[2024-01-03T19:08:40.846891198Z] [1|1] [info] NGEN is disabled.
[2024-01-03T19:08:40.847020103Z] [1|13] [info] Initializing ReJIT request thread.
[2024-01-03T19:08:40.847348038Z] [1|1] [info] Profiler filepath: /app/OpenTelemetry.AutoInstrumentation.Native.so
[2024-01-03T19:08:40.847362824Z] [1|1] [info] Profiler attached.
[2024-01-03T19:08:40.856924455Z] [1|1] [info] COR library: System.Private.CoreLib 8.0.0
[2024-01-03T19:08:41.391872308Z] [1|1] [info] ModuleLoadFinished: OpenTelemetry.AutoInstrumentation.Loader loaded into AppDomain 93981679129920 [clrhost]
[2024-01-03T19:08:41.458056240Z] [1|1] [info] Rewriting PInvokes to native: /app/OpenTelemetry.AutoInstrumentation.Native.so
[2024-01-03T19:08:41.458201997Z] [1|1] [info] Rewriting pinvoke for: AddInstrumentations
[2024-01-03T19:08:41.458220634Z] [1|1] [info] Rewriting pinvoke for: AddDerivedInstrumentations
[2024-01-03T19:08:41.984834060Z] [1|1] [info] AddInstrumentations: received id: FFAFA5168C4F4718B40CA8788875C2DA from managed side with 17 integrations.
[2024-01-03T19:08:41.984945482Z] [1|1] [info] Total number of modules to analyze: 23
[2024-01-03T19:08:41.985218818Z] [1|1] [info] InternalAddInstrumentation: Total integrations in profiler: 17
[2024-01-03T19:08:42.016708318Z] [1|13] [info] ReJIT handler stored metadata for 139748277831896 Microsoft.Extensions.Logging AppDomain 93981679129920 clrhost
[2024-01-03T19:08:42.016797738Z] [1|13] [info] Request ReJIT done for 1 methods
[2024-01-03T19:08:42.020218403Z] [1|1] [info] *** CallTarget_RewriterCallback() Finished: Microsoft.Extensions.Logging.LoggingBuilder..ctor() [IsVoid=1, IsStatic=0, IntegrationType=OpenTelemetry.AutoInstrumentation.Instrumentations.Logger.LoggingBuilderIntegration, Arguments=1]
```

### PR DESCRIPTION
**Pull request by Kielek**
## Why

Fixes #3212
Handles changes from https://github.com/open-telemetry/opentelemetry-dotnet/pull/3391

## What

Fixes AspNetCore Instrumentation for traces for .NET7.

It is now producing `Microsoft.AspNetCore` as it should.
What is more `OpenTelemetry.Instrumentation.AspNetCore` and its dependencies is loaded from additional store. It has to be loaded with dedicated-compiled version for each .NET version.

## Tests

Modified CI.

## Checklist

- [ ] `CHANGELOG.md` is updated.
- ~~[ ] Documentation is updated.~~
- [x] ~~New~~ features are covered by tests.

### CODE REVIEW

**Comment by Kielek** [comment]
As a follow up for this PR we should revisit strategy for loading other `OpenTelemetry.*` libraries and moved all of them to additional store.
I think that we can do it in the next release as we are not aware about any other issues.

File: test/IntegrationTests/BuildTests.DistributionStructure_alpine-linux.verified.txt [hunk 1]
```diff
@@ -71,6 +71,9 @@
   /store/x64/net7.0/microsoft.extensions.options.configurationextensions/8.0.0/lib/net8.0/Microsoft.Extensions.Options.ConfigurationExtensions.dll,
   /store/x64/net7.0/microsoft.extensions.options/8.0.0/lib/net8.0/Microsoft.Extensions.Options.dll,
   /store/x64/net7.0/microsoft.extensions.primitives/8.0.0/lib/net8.0/Microsoft.Extensions.Primitives.dll,
+  /store/x64/net7.0/opentelemetry.api.providerbuilderextensions/1.7.0/lib/net6.0/OpenTelemetry.Api.ProviderBuilderExtensions.dll,
```
Comment (pjanotti) [inline test/IntegrationTests/BuildTests.DistributionStructure_alpine-linux.verified.txt:74]: Do you understand why this assembly shows up under `net6.0` here, but, under `net8.0` on the line above for `net6.0` apps?

**Review by pjanotti — APPROVED** [review]
👍🏼 LGTM, one small Q about the `lib/<tfm >` for one assembly in the store for `net7.0` apps.

File: src/OpenTelemetry.AutoInstrumentation.AdditionalDeps/Directory.Build.props [hunk 2]
```diff
@@ -3,5 +3,6 @@
     <PackageReference Include="System.Diagnostics.DiagnosticSource" />
     <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" />
     <PackageReference Include="Microsoft.Extensions.Logging.Configuration" />
+    <PackageReference Include="OpenTelemetry.Instrumentation.AspNetCore" />
```
[the remark this answers is not in the document]
Comment (Kielek) [reply src/OpenTelemetry.AutoInstrumentation.AdditionalDeps/Directory.Build.props:6]: Do you see better idea to load dependecy targeted fo .NET7 in net7 runtime?
Based on what I see in store chnages, there is no new libraries included into it. I will double check it tommorow.

File: test/IntegrationTests/HttpTests.cs [hunk 3]
```diff
@@ -37,7 +37,11 @@ public void SubmitTraces(string propagators)
             return true;
         });
         Span? serverSpan = null;
+#if NET7_0_OR_GREATER
+        collector.Expect("Microsoft.AspNetCore", span =>
```
[the remark this answers is not in the document]
Comment (Kielek) [reply test/IntegrationTests/HttpTests.cs:41]: It is not "our" decision. The difference comes from the (semi)native support for ASP.NET Core instrumentation in net7+ and from  "OpenTelemetry.Instrumentation.AspNetCore" package.

I do not see possibility to have the same source for both versions. BTW same case is for http(client) instrumentation (already covered),

File: test/IntegrationTests/BuildTests.DistributionStructure_alpine-linux.verified.txt (hunk 1 above)
Comment (Kielek) [reply test/IntegrationTests/BuildTests.DistributionStructure_alpine-linux.verified.txt]: Missing roll forward - see d223f5656e9d79a367ad77f3dc782244a42d867a

**Comment by Kielek** [comment]
Handled by #3246

**Pull request closed without merging**